File: python3/easy-solution/1342.py
"""
1342. Number of Steps to Reduce a Number to Zero
    Given an integer num, return the number of steps to reduce it to zero.
    In one step, if the current number is even, you have to divide it by 2, otherwise, you have to subtract 1 from it.

    Example:
    Input: num = 14
    Output: 6
    Explanation: 
    Step 1) 14 is even; divide by 2 and obtain 7. 
    Step 2) 7 is odd; subtract 1 and obtain 6.
    Step 3) 6 is even; divide by 2 and obtain 3. 
    Step 4) 3 is odd; subtract 1 and obtain 2. 
    Step 5) 2 is even; divide by 2 and obtain 1. 
    Step 6) 1 is odd; subtract 1 and obtain 0.
"""

class Solution:
    def numberOfSteps(self, num: int) -> int:
        step = 0
        while num != 0:
            if num % 2 == 0:
                num /= 2
                step += 1
            else:
                num -= 1
                step += 1
        return step

File: JS/medium-solution/1342.js
/**
1432. Max Difference You Can Get From Changing an Integer
    You are given an integer num. You will apply the following steps to num two separate times:
    Pick a digit x (0 <= x <= 9).
    Pick another digit y (0 <= y <= 9). Note y can be equal to x.
    Replace all the occurrences of x in the decimal representation of num by y.
    Let a and b be the two results from applying the operation to num independently.
    Return the max difference between a and b.
    Note that neither a nor b may have any leading zeros, and must not be 0.

    Example :
    Input: num = 555
    Output: 888
    Explanation: The first time pick x = 5 and y = 9 and store the new integer in a.
    The second time pick x = 5 and y = 1 and store the new integer in b.
    We have now a = 999 and b = 111 and max difference = 888
 */



/**
 * @param {number} num
 * @return {number}
 */
var maxDiff = function(num) {
    let ten = Array(10).fill(1);
    for (let i = 1; i < 10; i++){
        ten[i] = ten[i - 1] * 10;
    }
    let value = 1;
    for (let i = 0; i < 10; i++){
        if (ten[i] > num){
            value = ten[i - 1];
            break;
        }
    }
    let a = -1, b = -1, x = 0, y = 0;
    const c = Math.floor(num / value) === 1 ? 0 : 1;

    while (value >= 1){
        const d = Math.floor(num / value);
        num %= value;
        if (a === -1 && d !== 9){
            a = d;
            x += 9 * value;
        } else if (d === a){
            x += 9 * value;
        } else {
            x += d * value;
        }
        if (b === -1 && d !== 0 && d !== 1){
            b = d;
            y += c * value;
        } else if (d === b){
            y += c * value;
        } else {
            y += d * value;
        }
        value = Math.floor(value / 10);
    }
    return x - y;
};